Tests/E2E/MacTest.cs

```diff
@@ -204,5 +204,139 @@ private async Task Test_OTAA_Unconfirmed_Send_And_Receive_C2D_Mac_CommandsImplAs
+            var linkADRCmdMessage = "LinkADRCmd mac command detected in upstream payload";
+            var changedDataRateMessage = $"\"datr\":\"SF9BW125\""; // DR3
+
+            // Sends 8x confirmed messages, stopping if C2D message is found and data rate is updated
+            for (var i = warmUpMessageCount + 1; i <= messagesToSend; ++i)
+            {
+                var msg = PayloadGenerator.Next().ToString(CultureInfo.InvariantCulture);
+                Log($"{device.DeviceID}: Sending unconfirmed '{msg}' {i}/{messagesToSend}");
+                await ArduinoDevice.transferPacketAsync(msg, 10);
+
+                await Task.Delay(Constants.DELAY_BETWEEN_MESSAGES);
+
+                // After transferPacket: Expectation from serial
+                // +MSG: Done
+                await AssertUtils.ContainsWithRetriesAsync("+MSG: Done", ArduinoDevice.SerialLogs);
+
+                // check if C2D message was found
+                // 0000000000000032: cloud to device MAC command LinkADRCmd received Type: LinkADRCmd
+                var C2DSearchResults = await TestFixtureCi.SearchNetworkServerModuleAsync(
+                    (messageBody) =>
+                    {
+                        return messageBody.StartsWith(c2dLogMessage, StringComparison.OrdinalIgnoreCase);
+                    },
+                    new SearchLogOptions
+                    {
+                        Description = c2dLogMessage,
+                        MaxAttempts = 1
+                    });
+
+                // check if LinkADRCmd MAC Command was detected
+                // LinkADRCmd mac command detected in upstream payload
+                var linkADRCmdSearchResults = await TestFixtureCi.SearchNetworkServerModuleAsync(
+                    (messageBody) =>
+                    {
+                        return messageBody.Contains(linkADRCmdMessage, StringComparison.OrdinalIgnoreCase);
+                    },
+                    new SearchLogOptions
+                    {
+                        Description = linkADRCmdMessage,
+                        MaxAttempts = 1
+                    });
+
+                // check if the data rate was changed to DR3
+                // 'datr':'SF9BW125'
+                var changedDataRateSearchResults = await TestFixtureCi.SearchNetworkServerModuleAsync(
+                    (messageBody) =>
+                    {
+                        return messageBody.Contains(changedDataRateMessage, StringComparison.OrdinalIgnoreCase);
+                    },
+                    new SearchLogOptions
+                    {
+                        Description = changedDataRateMessage,
+                        MaxAttempts = 1
+                    });
+
+                if (C2DSearchResults.Found)
+                {
+                    foundC2DMessageCount++;
+                    Log($"{device.DeviceID}: Found C2D message in log (after sending {i}/{messagesToSend}) {foundC2DMessageCount} times");
```

Comment:
Nice suggestion. I updated it in 56f8b527c2ca091d25bee732bf5a5df0cb38e8ac and also changed the counters to be bools instead since we're not counting the message occurrences anymore.